### DISCUSSION TITLE
How to remove thick long lines in image?

### DISCUSSION THREAD

**Discussion by nissansz** [python-pillow/Pillow#7539] [Q&A]
![25](https://github.com/python-pillow/Pillow/assets/41010669/b3057348-942c-43ec-b519-12cb64422f57)

**Comment by radarhere** [comment]
When you say 'remove', do you mean 'trim so the image is smaller',
or do you mean 'replace with white'?

And you're interested in affecting the top, left and bottom lines? Not just all red pixels?

**Comment by nissansz** [comment]
Replacing with white is fine.
Line colors may be different fron black, blue, etc. 
You can use binary image to locate lines and then replace

**Reply by radarhere** [reply]
> You can use binary image to locate lines and then replace

```python
from PIL import Image
im = Image.open("input.png")
im.convert("1").save("1.png")
```
gives
![1](https://github.com/python-pillow/Pillow/assets/3112309/c25f25b9-a831-4d80-90c8-1f36b0676a03)
I don't see how that idea helps.

**Comment by radarhere** [comment]
```python
from PIL import Image, ImageDraw
with Image.open("input.png") as im:
    processed = []
    rects = []
    x, y = 0, 0
    while y < im.height:
        while x < im.width:
            def matched(x3, y3):
                r, g, b, a = im.getpixel((x3, y3))
                if (r > 40 or g > 40 or b > 40) and (r < 240 or g < 240 or b < 240) and a == 255:
                    # Opaque and not black or white
                    return True
            if matched(x, y):
                x1, y1, x2, y2 = x, y, x, y
                
                # Search right
                while matched(x2, y2):
                    x2 += 1
                x2 -= 1
                
                # Search up
                while True:
                    lineMatches = True
                    for x3 in range(x1, x2+1):
                        if not matched(x3, y1):
                            lineMatches = False
                            break
                    if not lineMatches:
                        y1 += 1
                        break
                    y1 -= 1
                
                # Search down
                while True:
                    lineMatches = True
                    for x3 in range(x1, x2+1):
                        if not matched(x3, y2):
                            lineMatches = False
                            break
                    if not lineMatches:
                        y2 -= 1
                        break
                    y2 += 1
                
                if x2 - x1 > 5 and y2 - y1 > 5 and [x1, y1, x2, y2] not in rects:
                    rects.append([x1, y1, x2, y2])
                x = x2
            x += 1
        x = 0
        y += 1
    d = ImageDraw.Draw(im)
    for rect in rects:
        d.rectangle(rect, "#fff")
    im.save("out.png")
```
gives
![out](https://github.com/python-pillow/Pillow/assets/3112309/33a7c58f-b651-414b-9c09-1a9746060fe2)

**Comment by nissansz** [comment]
Thank you. When I try another image, it throws an error below.
```
    r, g, b, a = im.getpixel((x3, y3))
ValueError: not enough values to unpack (expected 4, got 3)
```
![27](https://github.com/python-pillow/Pillow/assets/41010669/168def93-d808-4753-bcf5-ddad5b7962df)

**Reply by radarhere** [reply]
Your first image was RGBA. This image is RGB, and so doesn't return an alpha value from `getpixel()`.

You can just insert `im = im.convert("RGBA")` to fix it.

**Comment by nissansz** [comment]
Thank you.
Some lines are not removed, and some words are damaged.
![image](https://github.com/python-pillow/Pillow/assets/41010669/03dc2986-b5fa-4875-8ed0-a3971bba42f4)

**Reply by radarhere** [reply]
Would it be possible for you to look at the code that I wrote for the first image, consider the approach that I've taken, and adjust the code to fit the second image?

**Comment by nissansz** [comment]
It is difficult to adjust.

                # if x2 - x1 > 5 and y2 - y1 > 5 and [x1, y1, x2, y2] not in rects:
                if x2 - x1 > 30 and y2 - y1 > 30 and [x1, y1, x2, y2] not in rects:

gives
![image](https://github.com/python-pillow/Pillow/assets/41010669/472610e8-fd18-41d1-bc16-6260052f64ae)

**Reply by radarhere** [reply]
Could you describe which lines should be removed?

There is
- a thick down arrow
- thin down arrows
- a thin arrow going right
- rectangle outlines
- lines between the rectangles
- two black lines in the middle of colours at the top

**Comment by nissansz** [comment]
I want to remove all long lines which are not a part of text

**Reply by radarhere** [reply]
As I [have](https://github.com/python-pillow/Pillow/issues/7376#issuecomment-1706317192) [said](https://github.com/python-pillow/Pillow/issues/7394#issuecomment-1718363072), Pillow can't detect text in an image. Therefore, it also cannot ignore text in an image.

If you would like to remove all lines that are not inside rectangles, that seems more feasible.

**Comment by nissansz** [comment]
Ok. Thank you.
